### PR TITLE
Movespeed modifiers, take two

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request

Refactors movespeed modifiers to be datum/movespeed_modifier(s) rather than lists.
adds a new proc for adding/updating variable movespeed modifiers as opposed to static movespeed modifiers which are now cached for performance/efficiency

## Why It's Good For The Game

datums can be cached (well lists technically can but no one's that crazy to do it with this) and are generally just better, don't know why I didn't just do this initially tbh

## Changelog
:cl:
refactor: movespeed modifiers are now datums and not lists and support global caching.
/:cl:

### CODE REVIEW

File: code/datums/components/mood.dm [hunk 1]
```diff
@@ -208,30 +208,43 @@
...
+			master.remove_movespeed_modifier(MOVESPEED_ID_SANITY)
 			sanity_level = 2
 		if(SANITY_GREAT+1 to INFINITY)
 			setInsanityEffect(0)
-			master.remove_movespeed_modifier(MOVESPEED_ID_SANITY, TRUE)
+			master.remove_movespeed_modifier(MOVESPEED_ID_SANITY)
 			sanity_level = 1
 	update_mood_icon()
 
+/datum/movespeed_modifier/sanity
```
Comment (ninjanomnom) [inline code/datums/components/mood.dm]: Put all the movespeed mods in a module along with the code managing them

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/mood.dm (hunk 1 above)
Comment (kevinz000) [reply code/datums/components/mood.dm]: woops forgot that one and some 1 sec

**Comment by ShizCalev** [comment]
Guessing this is fixed since you touched that section of code?

https://github.com/tgstation/tgstation/issues/48631

**Comment by HomocidlePancakes** [comment]
Will this fix #48164?

**Comment by optimumtact** [comment]
seems like this is just extra init and compile time for no real gain

**Comment by ninjanomnom** [comment]
The problem this is solving is that managing movespeed mods is maintenance hell and adding new functionality to the system has been getting more painful as we keep adding things to it

**Comment by optimumtact** [comment]
what functionality, and what maintenance hell

**Comment by ninjanomnom** [comment]
There's a lot of interactions between different kinds of speed modifiers that need tweaks to implement them but changing how speed mods work requires dealing with how all the data about those speed mods are spread in all the places that add or remove them as proc arguments. Adding new speed modifiers is not very intuitive as well.

It's really annoying to deal with that and reading/changing the code interpreting it is also more of a pain than it needs to be. The usability and maintainability of the system as a whole can be improved a lot beyond what it currently is.

**Comment by optimumtact** [comment]
alright, sounds good to me

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 2]
```diff
@@ -0,0 +1,257 @@
+/**
+  * Movespeed modification datums.
+  */
+
+/datum/movespeed_modifier
+	/// Whether or not this is a variable modifier. Variable modifiers can NOT be ever auto-cached. ONLY CHECKED VIA INITIAL(), EFFECTIVELY READ ONLY (and for very good reason)
+	var/variable = FALSE
+
+	/// Unique ID. You can never have different modifications with the same ID
+	var/id = "ERROR"
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: leave this null so its more sensible to check for if it's set properly

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+	/// Movetypes this never applies to
+	var/blacklisted_movetypes = NONE
+
+	/// Other modification datums this conflicts with.
+	var/conflicts_with
+
+/*! How move speed for mobs works
+
+Move speed is now calculated by using a list of movespeed modifiers, which is a list itself (to avoid datum overhead)
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: You have a lot of lying comments around here

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+GLOBAL_LIST_EMPTY(movespeed_modification_cache)
+
+/// Grabs a STATIC MODIFIER datum from cache. YOU MUST NEVER EDIT THESE DATUMS, OR IT WILL AFFECT ANYTHING ELSE USING IT TOO!
+/proc/get_cached_movespeed_modifier(modtype)
+	if(!ispath(modtype, /datum/movespeed_modifier))
+		CRASH("[modtype] is not a movespeed modification typepath.")
+	var/datum/movespeed_modifier/M = modtype
+	if(initial(M.variable))
+		CRASH("[modtype] is a variable modifier, and can never be cached.")
+	return GLOB.movespeed_modification_cache[modtype] || ((GLOB.movespeed_modification_cache[modtype] = new modtype))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Why the double wrapping?

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+///Add a move speed modifier to a mob. If a variable subtype is passed in as the first argument, it will make a new datum. If ID conflicts, it will overwrite the old ID.
+/mob/proc/add_movespeed_modifier(datum/movespeed_modifier/type_or_datum, update = TRUE)
+	if(ispath(type_or_datum))
+		if(!initial(type_or_datum.variable))
+			type_or_datum = get_cached_movespeed_modifier(type_or_datum)
+		else
+			type_or_datum = new type_or_datum
+	if(!istype(type_or_datum))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Dont bother checking this

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+		CRASH("Invalid modification datum")
+	var/oldpriority
+	var/datum/movespeed_modifier/existing = LAZYACCESS(movespeed_modification, type_or_datum.id)
+	if(existing)
+		if(existing == type_or_datum)		//same thing don't need to touch
+			return TRUE
+		oldpriority = existing.priority
+		remove_movespeed_modifier(existing, FALSE)
+	LAZYSET(movespeed_modification, type_or_datum.id, type_or_datum)
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: You probably want a binary insert here like what timers use to stay sorted

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+	return TRUE
+
+///Remove a move speed modifier from a mob, whether static or variable.
+/mob/proc/remove_movespeed_modifier(datum/movespeed_modifier/type_id_datum, update = TRUE)
+	if(ispath(type_id_datum))
+		if(!initial(type_id_datum.variable))
+			type_id_datum = get_cached_movespeed_modifier(type_id_datum)
+		else
+			type_id_datum = initial(type_id_datum.id)
+	if(istype(type_id_datum))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: again dont bother checking, if the wrong type sneaks in let it runtime normaly

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+		type_id_datum = type_id_datum.id
+	if(!istext(type_id_datum))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: ditto

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+		CRASH("Invalid ID")
+	if(!LAZYACCESS(movespeed_modification, type_id_datum))
+		return FALSE
+	LAZYREMOVE(movespeed_modification, type_id_datum)
+	UNSETEMPTY(movespeed_modification)
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: LAZYREMOVE already does this

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	3. Add the datum if necessary using the regular add proc
+	4. If any of the rest of the args are not null (see: multiplicative slowdown), modify the datum
+	5. Update if necessary
+	*/
+	var/modified = FALSE
+	var/inject = FALSE
+	var/datum/movespeed_modifier/final
+	if(istext(type_id_datum))
+		final = LAZYACCESS(movespeed_modification, type_id_datum)
+		if(!istype(final))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]:
```suggestion
		if(!final)
```

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+			CRASH("Couldn't find existing modification when only provided an ID.")
+	else if(ispath(type_id_datum))
+		if(!initial(type_id_datum.variable))
+			CRASH("Not a variable modifier")
+		var/id = initial(type_id_datum.id)
+		final = LAZYACCESS(movespeed_modification, id)
+		if(!istype(final))
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]:
```suggestion
		if(!final)
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
The organization for the movespeed mod files also doesn't make too much sense to me but whatever.

File: code/modules/movespeed/_movespeed_modifier.dm (hunk 4 above)
Comment (kevinz000) [reply code/modules/movespeed/_movespeed_modifier.dm]: Micro optimization I learned when doing getflaticon, this **seems** to be faster than a var set behind an if block

Comment (ninjanomnom) [reply code/modules/movespeed/_movespeed_modifier.dm]: I'm sorry ***what***

I'll need to see profile data for that one

Comment (optimumtact) [reply code/modules/movespeed/_movespeed_modifier.dm]: I hate you both

Comment (kevinz000) [reply code/modules/movespeed/_movespeed_modifier.dm]: image
image

aight bet gamers

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 12]
```diff
@@ -79,17 +79,36 @@ GLOBAL_LIST_EMPTY(movespeed_modification_cache)
...
+			curr = movespeed_modification[id]
+			if(curr.priority <= type_or_datum.priority)
+				left = mid + 1
+			else
+				right = mid
+			mid = (left + right) >> 1
+		curr = movespeed_modification[mid]
+		mid = curr.priority > type_or_datum.priority? mid : mid + 1
+		movespeed_modification.Insert(mid, type_or_datum.id)
+		movespeed_modification[type_or_datum.id] = type_or_datum
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Use #48831 for this when it's merged

File: code/modules/antagonists/slaughter/slaughter.dm [hunk 13]
```diff
@@ -33,7 +33,7 @@
 	health = 200
 	healable = 0
 	environment_smash = ENVIRONMENT_SMASH_STRUCTURES
-	obj_damage = 50
+	obj_damage = 5
```
Comment (ninjanomnom) [inline code/modules/antagonists/slaughter/slaughter.dm]: ?

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+	LAZYSET(movespeed_modification, type_or_datum.id, type_or_datum)
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: By this point you know the list exists and this doesn't need to be lazy

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by optimumtact** [comment]
It would help if it compiled

**Comment by spookydonut** [comment]
```
[09:31:44] Runtime in _movespeed_modifier.dm,105: Cannot read "MOB_WALK_RUN".priority

  proc name: add movespeed modifier (/mob/proc/add_movespeed_modifier)

  src: Hattori (/mob/living/simple_animal/bot/medbot)

  src.loc: the floor (108,45,1) (/turf/open/floor/wood)

  call stack:

  Hattori (/mob/living/simple_animal/bot/medbot): add movespeed modifier(/datum/movespeed_modifier/mob_... (/datum/movespeed_modifier/mob_config_speedmod), 0)

  Hattori (/mob/living/simple_animal/bot/medbot): add or update variable movespeed modifier(/datum/movespeed_modifier/mob_... (/datum/movespeed_modifier/mob_config_speedmod), 1, 0)

  Hattori (/mob/living/simple_animal/bot/medbot): update config movespeed()

  Hattori (/mob/living/simple_animal/bot/medbot): update config movespeed()

  Hattori (/mob/living/simple_animal/bot/medbot): Initialize(1, null)

  Hattori (/mob/living/simple_animal/bot/medbot): Initialize(1, null)

  Hattori (/mob/living/simple_animal/bot/medbot): Initialize(1, null)

  Hattori (/mob/living/simple_animal/bot/medbot): Initialize(1, null)

  Hattori (/mob/living/simple_animal/bot/medbot): Initialize(1, null)

  Atoms (/datum/controller/subsystem/atoms): InitAtom(Hattori (/mob/living/simple_animal/bot/medbot), /list (/list))

  Atoms (/datum/controller/subsystem/atoms): InitializeAtoms(null)

  Atoms (/datum/controller/subsystem/atoms): Initialize(343049)

  Master (/datum/controller/master): Initialize(10, 0, 1)
```
etc

File: code/modules/antagonists/slaughter/slaughter.dm (hunk 13 above)
Comment (kevinz000) [reply code/modules/antagonists/slaughter/slaughter.dm]: *SCREAM 
probably pressed backspace

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+GLOBAL_LIST_EMPTY(movespeed_modification_cache)
+
+/// Grabs a STATIC MODIFIER datum from cache. YOU MUST NEVER EDIT THESE DATUMS, OR IT WILL AFFECT ANYTHING ELSE USING IT TOO!
+/proc/get_cached_movespeed_modifier(modtype)
+	if(!ispath(modtype, /datum/movespeed_modifier))
+		CRASH("[modtype] is not a movespeed modification typepath.")
+	var/datum/movespeed_modifier/M = modtype
+	if(initial(M.variable))
+		CRASH("[modtype] is a variable modifier, and can never be cached.")
+	return GLOB.movespeed_modification_cache[modtype] || (GLOB.movespeed_modification_cache[modtype] = new modtype)
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Don't this in one statement, assigning in the return is gross

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+			type_or_datum = get_cached_movespeed_modifier(type_or_datum)
+		else
+			type_or_datum = new type_or_datum
+	var/datum/movespeed_modifier/existing = LAZYACCESS(movespeed_modification, type_or_datum.id)
+	if(existing)
+		if(existing == type_or_datum)		//same thing don't need to touch
+			return TRUE
+		remove_movespeed_modifier(existing, FALSE)
+	if(length(movespeed_modification))
+		BINARY_INSERT(type_or_datum.id, movespeed_modification, datum/movespeed_modifier, type_or_datum, priority, __BIN_LIST[__BIN_LIST[__BIN_MID]])
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: We have defines for the last argument here

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by kevinz000** [comment]
help bot bullying me @ninjanomnom

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+can next move
+
+Key procs
+* [add_movespeed_modifier](mob.html#proc/add_movespeed_modifier)
+* [remove_movespeed_modifier](mob.html#proc/remove_movespeed_modifier)
+* [has_movespeed_modifier](mob.html#proc/has_movespeed_modifier)
+* [update_movespeed](mob.html#proc/update_movespeed)
+*/
+
+//ANY ADD/REMOVE DONE IN UPDATE_MOVESPEED MUST HAVE THE UPDATE ARGUMENT SET AS FALSE!
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Why is this here instead of somewhere relevant to the comment

File: code/modules/movespeed/_movespeed_modifier.dm [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+		return FALSE
+	LAZYREMOVE(movespeed_modification, type_id_datum)
+	if(update)
+		update_movespeed(FALSE)
+	return TRUE
+
+/// Used for variable slowdowns like hunger/health loss/etc, works somewhat like the old list-based modification adds. Returns the modifier datum if successful
+/mob/proc/add_or_update_variable_movespeed_modifier(datum/movespeed_modifier/type_id_datum, update = TRUE, multiplicative_slowdown)
+	/*
+	How this SHOULD work is:
```
Comment (ninjanomnom) [inline code/modules/movespeed/_movespeed_modifier.dm]: Why isn't this in the autodoc comment for the proc

File: code/modules/movespeed/modifiers/reagent.dm [hunk 19]
```diff
@@ -0,0 +1,42 @@
+/datum/movespeed_modifier/reagent
+	blacklisted_movetypes = (FLYING|FLOATING)
+
+/datum/movespeed_modifier/reagent/stimulants
+	id = "stimulants_reagent"
```
Comment (ninjanomnom) [inline code/modules/movespeed/modifiers/reagent.dm]: Can't you just use the type for most things and only have a text id when multiple types need to share one id?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
Seems mostly fine on the code side, I'll testmerge this later

**Comment by kevinz000** [comment]
working on it 1 sec

**Comment by ShizCalev** [comment]
does this fix #48631

**Comment by ninjanomnom** [comment]
I doubt it

**Comment by kevinz000** [comment]
no but it'll make it less cancer to fix

File: code/modules/movespeed/_movespeed_modifier.dm (hunk 18 above)
Comment (kevinz000) [reply code/modules/movespeed/_movespeed_modifier.dm]: fixed

**Comment by kevinz000** [comment]
@ninjanomnom

File: code/modules/mob/living/living_movement.dm [hunk 20]
```diff
@@ -26,38 +26,31 @@
...
-		mod = CONFIG_GET(number/movedelay/run_delay)
-	if(!isnum(mod))
-		mod = 1
-	add_movespeed_modifier(MOVESPEED_ID_MOB_WALK_RUN_CONFIG_SPEED, TRUE, 100, override = TRUE, multiplicative_slowdown = mod)
+	add_movespeed_modifier((m_intent == MOVE_INTENT_WALK)? /datum/movespeed_modifier/config_walk_run/walk : /datum/movespeed_modifier/config_walk_run/run)
 
 /mob/living/proc/update_turf_movespeed(turf/open/T)
-	if(isopenturf(T))
-		add_movespeed_modifier(MOVESPEED_ID_LIVING_TURF_SPEEDMOD, update=TRUE, priority=100, override=TRUE, multiplicative_slowdown=T.slowdown, movetypes=GROUND, blacklisted_movetypes=(FLYING|FLOATING))
+	if(istype(T))
```
Comment (Rohesie) [inline code/modules/mob/living/living_movement.dm]:
```suggestion
	if(isopenturf(T))
```
It's better if the check remains as is and `turf/open/T` gets changed to something like `atom/T` as else it's being misleading. The proc calling this uses `loc` on `Moved()`, and that can be any `atom`.

File: code/modules/mob/mob.dm [hunk 21]
```diff
@@ -1234,16 +1234,22 @@
...
+		remove_movespeed_modifier(MOVESPEED_ID_MOB_GRAB_STATE)
 	else
-		add_movespeed_modifier(MOVESPEED_ID_MOB_GRAB_STATE, update=TRUE, priority=100, override=TRUE, multiplicative_slowdown=grab_state*3, blacklisted_movetypes=FLOATING)
+		switch(grab_state)
+			if(GRAB_AGGRESSIVE)
+				add_movespeed_modifier(/datum/movespeed_modifier/grab_slowdown/aggressive)
+			if(GRAB_NECK)
+				add_movespeed_modifier(/datum/movespeed_modifier/grab_slowdown/neck)
+			if(GRAB_KILL)
+				add_movespeed_modifier(/datum/movespeed_modifier/grab_slowdown/kill)
```
Comment (Rohesie) [inline code/modules/mob/mob.dm]: Why not a single `switch`?

**Comment by DomitiusKnack** [comment]
Breaks morphine. Maximum speed even when crawling.
